pkg/applicationserver/applicationserver.go

```diff
@@ -626,31 +826,31 @@ func (as *ApplicationServer) handleUplink(ctx context.Context, ids ttnpb.EndDevi
 					dev.DevAddr = ids.DevAddr
 					mask = append(mask, "session", "ids.dev_addr")
 					logger.Debug("Restored session")
-				}
-				dev.PendingSession = nil
-				mask = append(mask, "pending_session")
 
-				// At this point, the application downlink queue in the Network Server is invalid; recalculation is necessary.
-				// Next AFCntDown 1 is assumed. If this is a LoRaWAN 1.0.x end device and the Network Server sent MAC layer
-				// downlink already, the Network Server will trigger the DownlinkQueueInvalidated event. Therefore, this
-				// recalculation may result in another recalculation.
-				client := ttnpb.NewAsNsClient(link.conn)
-				res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
-				if err != nil {
-					log.WithError(err).Warn("Failed to list downlink queue for recalculation; clearing the downlink queue")
-					req := &ttnpb.DownlinkQueueRequest{
-						EndDeviceIdentifiers: ids,
-					}
-					_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
+					// At this point, the application downlink queue in the Network Server is invalid; recalculation is necessary.
+					// Next AFCntDown 1 is assumed. If this is a LoRaWAN 1.0.x end device and the Network Server sent MAC layer
+					// downlink already, the Network Server will trigger the DownlinkQueueInvalidated event. Therefore, this
+					// recalculation may result in another recalculation.
+					client := ttnpb.NewAsNsClient(link.conn)
```

Comment:
AS does not need to do anything here, because NS simply discards the pending session, and the current queue is already there.

Reply:
I disagree. Unless I'm missing something very obvious, here the following conditions hold:
1. `dev.Session == nil || !bytes.Equal(dev.Session.SessionKeyID, uplink.SessionKeyID)` (line 807).
2. `!(dev.PendingSession != nil && bytes.Equal(dev.PendingSession.SessionKeyID, uplink.SessionKeyID))` (line 808).

Due to 1, the downlink queue in NS for `dev.Session` cannot be valid, since it uses a different session key.

Reply:
Assume device has current queue `CQ` and pending queue `PQ` (either of them may be `nil`)
1. Device sends a join-request, NS receives a join-accept from JS
2. AS receives `CQ` if not `nil` and `PQ` otherwise, reencrypts using the new key. Let's call the resulting queue `PQ'`. AS replaces the device queue, which is now `CQ ++ PQ'`
3. NS may arbitrary mutate `CQ` by scheduling application downlinks (not implemented currently). `PQ'` does not get mutated. Let's call the new, possibly mutated, current queue `CQ'`
4. Device sends an uplink using either the key used for `CQ'` or `PQ'`
5. NS discards the queue for which the key does not match and takes either `CQ'` or `PQ'` as the current queue according to that.
5. AS discards one of the sessions. There's nothing else to do, because both `CQ'` and `PQ'` are already valid.

Reply:
This use case, in fact, is exactly the reason for the issue we're closing here (https://github.com/TheThingsNetwork/lorawan-stack/issues/1954).

> #### Why do we need this?
> To avoid hitting AS to recalculate the downlink queue when the end device > confirms the session. Having a downlink queue of both the current session > and the pending session in NS allows for downlink items to be available immediately.